File: core/src/net/brenig/pixelescape/lib/LogHelper.java
package net.brenig.pixelescape.lib;

import com.badlogic.gdx.Application;
import com.badlogic.gdx.Gdx;
import net.brenig.pixelescape.game.data.GameDebugSettings;

public final class LogHelper {

	public enum LogLevel {
		DEBUG("DEBUG", Application.LOG_DEBUG, Gdx.app::debug, Gdx.app::debug),
		INFO("INFO", Application.LOG_INFO, Gdx.app::log, Gdx.app::log),
		WARNING("WARN", Application.LOG_ERROR, Gdx.app::error, Gdx.app::error),
		ERROR("ERROR", Application.LOG_ERROR, Gdx.app::error, Gdx.app::error);

		String tag;
		int level;
		NormalLogger normalLogger;
		ExceptionLogger exceptionLogger;

		LogLevel(String tag, int level, NormalLogger normalLogger, ExceptionLogger exceptionLogger) {
			this.tag = tag;
			this.level = level;
			this.normalLogger = normalLogger;
			this.exceptionLogger = exceptionLogger;
		}
	}

	public interface NormalLogger {
		void log(String tag, String msg);
	}
	public interface ExceptionLogger {
		void log(String tag, String msg, Throwable t);
	}

	private static final String LOG_TAG_BRACKET_OPEN = "[";
	private static final String LOG_TAG_BRACKET_CLOSE = "]";
	private static final String LOG_TAG_BRACKET_SEPARATE = " | ";
	private static final String LOG_LEVEL_BRACKET_OPEN = "[";
	private static final String LOG_LEVEL_BRACKET_CLOSE = "]";

	private static final String LOG_TAG_NAME = "PixelEscape";

	public static void log(String msg) {
		log(LogLevel.INFO, null, msg, null);
	}

	public static void log(String tag, String msg) {
		log(LogLevel.INFO, tag, msg, null);
	}

	public static void log(LogLevel level, String tag, String msg) {
		log(level, tag, msg, null);
	}

	public static void log(LogLevel level, String tag, String msg, Throwable t) {
		StringBuilder builder = new StringBuilder();
		//level
		builder.append(LOG_LEVEL_BRACKET_OPEN).append(level.tag).append(LOG_LEVEL_BRACKET_CLOSE);
		//pixelescape tag
		builder.append(LOG_TAG_BRACKET_OPEN + LOG_TAG_NAME);
		//tag
		if (tag != null && !(tag.length() <= 0)) {
			builder.append(LOG_TAG_BRACKET_SEPARATE).append(tag);
		}
		builder.append(LOG_TAG_BRACKET_CLOSE);

		if (t != null) {
			level.exceptionLogger.log(builder.toString(), msg, t);
		} else {
			level.normalLogger.log(builder.toString(), msg);
		}
	}

	public static void debug(String msg) {
		debug(null, msg, null);
	}

	public static void debug(String tag, String msg) {
		debug(tag, msg, null);
	}

	public static void debug(String tag, String msg, Throwable t) {
		if (GameDebugSettings.get("DEBUG_LOGGING")) {
			log(LogLevel.DEBUG, tag, msg, t);
		}
	}

	public static void error(String msg) {
		log(LogLevel.ERROR, null, msg, null);
	}

	public static void error(String tag, String msg) {
		log(LogLevel.ERROR, tag, msg, null);
	}

	public static void error(String tag, String msg, Throwable t) {
		log(LogLevel.ERROR, tag, msg, t);
	}

	public static void warn(String msg) {
		log(LogLevel.WARNING, null, msg, null);
	}

	public static void warn(String tag, String msg) {
		log(LogLevel.WARNING, tag, msg, null);
	}

	public static void warn(String tag, String msg, Throwable t) {
		log(LogLevel.WARNING, tag, msg, t);
	}

	public static void info(String msg) {
		log(LogLevel.INFO, null, msg, null);
	}

	public static void info(String tag, String msg) {
		log(LogLevel.INFO, tag, msg, null);
	}

	public static void info(String tag, String msg, Throwable t) {
		log(LogLevel.INFO, tag, msg, t);
	}

	public static void setGDXLogLevel(int level) {
		Gdx.app.setLogLevel(level);
	}

	private LogHelper() {}
}

File: core/src/net/brenig/pixelescape/lib/LogHelper.kt
package net.brenig.pixelescape.lib

import com.badlogic.gdx.Application
import com.badlogic.gdx.Gdx
import net.brenig.pixelescape.game.data.GameDebugSettings

private const val LOG_TAG_BRACKET_OPEN = "["
private const val LOG_TAG_BRACKET_CLOSE = "]"
private const val LOG_TAG_BRACKET_SEPARATE = " | "
private const val LOG_LEVEL_BRACKET_OPEN = "["
private const val LOG_LEVEL_BRACKET_CLOSE = "]"

private const val LOG_TAG_NAME = "PixelEscape"

enum class LogLevel(internal var tag: String, internal var level: Int) {
    DEBUG("DEBUG", Application.LOG_DEBUG),
    INFO("INFO", Application.LOG_INFO),
    WARNING("WARN", Application.LOG_ERROR),
    ERROR("ERROR", Application.LOG_ERROR)
}

fun log(msg: String) {
    log(LogLevel.INFO, null, msg, null)
}

fun log(tag: String, msg: String) {
    log(LogLevel.INFO, tag, msg, null)
}

fun log(level: LogLevel, tag: String?, msg: String, t: Throwable? = null) {
    val builder = StringBuilder()
    //level
    builder.append(LOG_LEVEL_BRACKET_OPEN).append(level.tag).append(LOG_LEVEL_BRACKET_CLOSE)
    //pixelescape tag
    builder.append(LOG_TAG_BRACKET_OPEN + LOG_TAG_NAME)
    //tag
    if (!tag.isNullOrEmpty()) {
        builder.append(LOG_TAG_BRACKET_SEPARATE).append(tag)
    }
    builder.append(LOG_TAG_BRACKET_CLOSE)

    if (t != null) {
        when (level) {
            LogLevel.DEBUG -> Gdx.app.debug(builder.toString(), msg, t)
            LogLevel.INFO -> Gdx.app.log(builder.toString(), msg, t)
            LogLevel.WARNING, LogLevel.ERROR -> Gdx.app.error(builder.toString(), msg, t)
        }
    } else {
        when (level) {
            LogLevel.DEBUG -> Gdx.app.debug(builder.toString(), msg)
            LogLevel.INFO -> Gdx.app.log(builder.toString(), msg)
            LogLevel.WARNING, LogLevel.ERROR -> Gdx.app.error(builder.toString(), msg)
        }
    }
}

fun debug(msg: String) {
    debug(null, msg, null)
}

@JvmOverloads
fun debug(tag: String?, msg: String, t: Throwable? = null) {
    if (GameDebugSettings["DEBUG_LOGGING"]) {
        log(LogLevel.DEBUG, tag, msg, t)
    }
}

fun error(msg: String) {
    log(LogLevel.ERROR, null, msg, null)
}

fun error(tag: String?, msg: String, t: Throwable? = null) {
    log(LogLevel.ERROR, tag, msg, t)
}

fun warn(msg: String) {
    log(LogLevel.WARNING, null, msg, null)
}

fun warn(tag: String?, msg: String, t: Throwable? = null) {
    log(LogLevel.WARNING, tag, msg, t)
}

fun info(msg: String) {
    log(LogLevel.INFO, null, msg, null)
}

fun info(tag: String?, msg: String, t: Throwable? = null) {
    log(LogLevel.INFO, tag, msg, t)
}

fun setGDXLogLevel(level: Int) {
    Gdx.app.logLevel = level
}

fun setGDXLogLevel(level: LogLevel) {
    Gdx.app.logLevel = level.level
}